pkg/util/conn_pool.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2020 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ceph/go-ceph/rados"
+)
+
+type connEntry struct {
+	conn     *rados.Conn
+	lastUsed time.Time
+	unique   string
+	users    int
+}
+
+type ConnPool struct {
+	// interval to run the garbage collector
+	interval time.Duration
+	// timeout for a connEntry to get garbage collected
+	expiry time.Duration
+	// Timer used to schedule calls to the garbage collector
+	timer *time.Timer
+	// Mutex for loading and touching connEntry's from the conns Map
+	lock *sync.RWMutex
+	// all connEntry's in this pool
+	conns map[string]*connEntry
+}
+
+// Create a new ConnPool instance and start the garbage collector running
+// every @interval.
+func NewConnPool(interval, expiry time.Duration) *ConnPool {
+	cp := ConnPool{
+		interval: interval,
+		expiry:   expiry,
+		lock:     &sync.RWMutex{},
+		conns:    make(map[string]*connEntry),
+	}
+	cp.timer = time.AfterFunc(interval, cp.gc)
+
+	return &cp
+}
+
+// loop through all cp.conns and destroy objects that have not been used for cp.expiry.
+func (cp *ConnPool) gc() {
+	cp.lock.Lock()
+	defer cp.lock.Unlock()
+
+	now := time.Now()
+	for key, ce := range cp.conns {
+		if ce.users == 0 && (now.Sub(ce.lastUsed)) > cp.expiry {
+			ce.destroy()
+			delete(cp.conns, key)
+		}
+	}
+
+	// schedule the next gc() run
+	cp.timer.Reset(cp.interval)
+}
+
+// Stop the garbage collector and destroy all connections in the pool.
+func (cp *ConnPool) Destroy() {
+	cp.timer.Stop()
+	// wait until gc() has finished, in case it is running
+	cp.lock.Lock()
+	defer cp.lock.Unlock()
+
+	for key, ce := range cp.conns {
+		if ce.users != 0 {
+			// should never happen
+			// TODO: return error, or panic()?
```

Comment:
If the only way for this to happen is a logic bug in this code, I'd suggest to panic with a very clear message.